diff -r 33c333ec5ad2 dyncall/dyncall.h
--- a/dyncall/dyncall.h	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall.h	Sat Mar 12 10:54:27 2016 +0300
@@ -97,6 +97,8 @@
 
 DC_API void       dcMode          (DCCallVM* vm, DCint mode);
 
+DC_API void       dcRetStruct     (DCCallVM* vm, DCstruct* s, DCpointer returnValue);
+
 DC_API void       dcArgBool       (DCCallVM* vm, DCbool     value);
 DC_API void       dcArgChar       (DCCallVM* vm, DCchar     value);
 DC_API void       dcArgShort      (DCCallVM* vm, DCshort    value);
@@ -118,7 +120,7 @@
 DC_API DCfloat    dcCallFloat     (DCCallVM* vm, DCpointer funcptr);
 DC_API DCdouble   dcCallDouble    (DCCallVM* vm, DCpointer funcptr);
 DC_API DCpointer  dcCallPointer   (DCCallVM* vm, DCpointer funcptr);
-DC_API void       dcCallStruct    (DCCallVM* vm, DCpointer funcptr, DCstruct* s, DCpointer returnValue);
+DC_API void       dcCallStruct    (DCCallVM* vm, DCpointer funcptr);
 
 DC_API DCint      dcGetError      (DCCallVM* vm);
 
diff -r 33c333ec5ad2 dyncall/dyncall_api.c
--- a/dyncall/dyncall_api.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_api.c	Sat Mar 12 10:54:27 2016 +0300
@@ -46,6 +46,11 @@
    * a dcMode(callvm, DC_CALL_C_ELLIPSIS_VARARGS) */
 }
 
+void dcRetStruct(DCCallVM* vm, DCstruct* s, DCpointer x) 
+{ 
+  vm->mVTpointer->retStruct(vm, s, x); 
+}
+
 void dcArgBool(DCCallVM* vm,DCbool x) 
 { 
   vm->mVTpointer->argBool(vm, x); 
@@ -147,9 +152,9 @@
   return vm->mVTpointer->callPointer(vm, funcptr); 
 }
 
-void dcCallStruct(DCCallVM* vm, DCpointer funcptr, DCstruct* s, DCpointer x) 
+void dcCallStruct(DCCallVM* vm, DCpointer funcptr) 
 { 
-  vm->mVTpointer->callStruct(vm, funcptr, s, x); 
+  vm->mVTpointer->callStruct(vm, funcptr); 
 }
 
 DCint dcGetError(DCCallVM *vm)
diff -r 33c333ec5ad2 dyncall/dyncall_callvm.h
--- a/dyncall/dyncall_callvm.h	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm.h	Sat Mar 12 10:54:27 2016 +0300
@@ -47,6 +47,7 @@
   void         (*free)           (DCCallVM* vm);
   void         (*reset)          (DCCallVM* vm);
   void         (*mode)           (DCCallVM* vm,DCint        mode);
+  void         (*retStruct)      (DCCallVM* vm,DCstruct* s, DCpointer returnValue);
   void         (*argBool)        (DCCallVM* vm,DCbool          B);
   void         (*argChar)        (DCCallVM* vm,DCchar          c);
   void         (*argShort)       (DCCallVM* vm,DCshort         s);
@@ -67,7 +68,7 @@
   DCfloat      (*callFloat)      (DCCallVM* vm,DCpointer funcptr);
   DCdouble     (*callDouble)     (DCCallVM* vm,DCpointer funcptr);
   DCpointer    (*callPointer)    (DCCallVM* vm,DCpointer funcptr);
-  void         (*callStruct)     (DCCallVM* vm,DCpointer funcptr,DCstruct* s, DCpointer returnValue);
+  void         (*callStruct)     (DCCallVM* vm,DCpointer funcptr);
 };
 
 typedef DCvoid       (DCvoidvmfunc)      (DCCallVM* vm,DCpointer funcptr); 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_arm32_arm.c
--- a/dyncall/dyncall_callvm_arm32_arm.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_arm32_arm.c	Sat Mar 12 10:54:27 2016 +0300
@@ -158,6 +158,7 @@
   &dc_callvm_free_arm32_arm
 , &dc_callvm_reset_arm32_arm
 , &dc_callvm_mode_arm32_arm
+, NULL /* retStruct */
 , &dc_callvm_argBool_arm32_arm
 , &dc_callvm_argChar_arm32_arm
 , &dc_callvm_argShort_arm32_arm 
@@ -187,6 +188,7 @@
   &dc_callvm_free_arm32_arm
 , &dc_callvm_reset_arm32_arm
 , &dc_callvm_mode_arm32_arm
+, NULL /* retStruct */
 , &dc_callvm_argBool_arm32_arm
 , &dc_callvm_argChar_arm32_arm
 , &dc_callvm_argShort_arm32_arm 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_arm64.c
--- a/dyncall/dyncall_callvm_arm64.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_arm64.c	Sat Mar 12 10:54:27 2016 +0300
@@ -117,6 +117,7 @@
   &deinit
 , &reset
 , &mode
+, NULL /* retStruct */
 , &a_bool
 , &a_char
 , &a_short 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_mips_eabi.c
--- a/dyncall/dyncall_callvm_mips_eabi.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_mips_eabi.c	Sat Mar 12 10:54:27 2016 +0300
@@ -147,6 +147,7 @@
   &dc_callvm_free_mips_eabi
 , &dc_callvm_reset_mips_eabi
 , &dc_callvm_mode_mips_eabi
+, NULL /* retStruct */
 , &dc_callvm_argBool_mips_eabi
 , &dc_callvm_argChar_mips_eabi
 , &dc_callvm_argShort_mips_eabi 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_mips_n32.c
--- a/dyncall/dyncall_callvm_mips_n32.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_mips_n32.c	Sat Mar 12 10:54:27 2016 +0300
@@ -185,6 +185,7 @@
   &dc_callvm_free_mips_n32
 , &dc_callvm_reset_mips_n32
 , &dc_callvm_mode_mips_n32
+, NULL /* retStruct */
 , &dc_callvm_argBool_mips_n32
 , &dc_callvm_argChar_mips_n32
 , &dc_callvm_argShort_mips_n32 
@@ -213,6 +214,7 @@
   &dc_callvm_free_mips_n32
 , &dc_callvm_reset_mips_n32
 , &dc_callvm_mode_mips_n32
+, NULL /* retStruct */
 , &dc_callvm_argBool_mips_n32
 , &dc_callvm_argChar_mips_n32
 , &dc_callvm_argShort_mips_n32
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_mips_n64.c
--- a/dyncall/dyncall_callvm_mips_n64.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_mips_n64.c	Sat Mar 12 10:54:27 2016 +0300
@@ -185,6 +185,7 @@
   &dc_callvm_free_mips_n64
 , &dc_callvm_reset_mips_n64
 , &dc_callvm_mode_mips_n64
+, NULL /* retStruct */
 , &dc_callvm_argBool_mips_n64
 , &dc_callvm_argChar_mips_n64
 , &dc_callvm_argShort_mips_n64 
@@ -213,6 +214,7 @@
   &dc_callvm_free_mips_n64
 , &dc_callvm_reset_mips_n64
 , &dc_callvm_mode_mips_n64
+, NULL /* retStruct */
 , &dc_callvm_argBool_mips_n64
 , &dc_callvm_argChar_mips_n64
 , &dc_callvm_argShort_mips_n64
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_mips_o32.c
--- a/dyncall/dyncall_callvm_mips_o32.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_mips_o32.c	Sat Mar 12 10:54:27 2016 +0300
@@ -200,6 +200,7 @@
   &dc_callvm_free_mips_o32
 , &dc_callvm_reset_mips_o32
 , &dc_callvm_mode_mips_o32
+, NULL /* retStruct */
 , &dc_callvm_argBool_mips_o32
 , &dc_callvm_argChar_mips_o32
 , &dc_callvm_argShort_mips_o32 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_ppc32.c
--- a/dyncall/dyncall_callvm_ppc32.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_ppc32.c	Sat Mar 12 10:54:27 2016 +0300
@@ -282,6 +282,7 @@
   &dc_callvm_free_ppc32
 , &dc_callvm_reset_ppc32
 , &dc_callvm_mode_ppc32
+, NULL /* retStruct */
 , &dc_callvm_argBool_ppc32
 , &dc_callvm_argChar_ppc32
 , &dc_callvm_argShort_ppc32 
@@ -310,6 +311,7 @@
   &dc_callvm_free_ppc32
 , &dc_callvm_reset_ppc32
 , &dc_callvm_mode_ppc32
+, NULL /* retStruct */
 , &dc_callvm_argBool_ppc32
 , &dc_callvm_argChar_ppc32
 , &dc_callvm_argShort_ppc32 
@@ -338,6 +340,7 @@
   &dc_callvm_free_ppc32
 , &dc_callvm_reset_ppc32
 , &dc_callvm_mode_ppc32
+, NULL /* retStruct */
 , &dc_callvm_argBool_ppc32
 , &dc_callvm_argChar_ppc32
 , &dc_callvm_argShort_ppc32 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_ppc64.c
--- a/dyncall/dyncall_callvm_ppc64.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_ppc64.c	Sat Mar 12 10:54:27 2016 +0300
@@ -224,6 +224,7 @@
   &dc_callvm_free_ppc64
 , &dc_callvm_reset_ppc64
 , &dc_callvm_mode_ppc64
+, NULL /* retStruct */
 , &dc_callvm_argBool_ppc64
 , &dc_callvm_argChar_ppc64
 , &dc_callvm_argShort_ppc64
@@ -253,6 +254,7 @@
   &dc_callvm_free_ppc64
 , &dc_callvm_reset_ppc64
 , &dc_callvm_mode_ppc64
+, NULL /* retStruct */
 , &dc_callvm_argBool_ppc64
 , &dc_callvm_argChar_ppc64
 , &dc_callvm_argShort_ppc64
@@ -282,6 +284,7 @@
   &dc_callvm_free_ppc64
 , &dc_callvm_reset_ppc64
 , &dc_callvm_mode_ppc64
+, NULL /* retStruct */
 , &dc_callvm_argBool_ppc64
 , &dc_callvm_argChar_ppc64
 , &dc_callvm_argShort_ppc64
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_sparc.c
--- a/dyncall/dyncall_callvm_sparc.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_sparc.c	Sat Mar 12 10:54:27 2016 +0300
@@ -125,6 +125,7 @@
   &dc_callvm_free_sparc, 
   &dc_callvm_reset_sparc, 
   &dc_callvm_mode_sparc, 
+  NULL /* retStruct */,
   &dc_callvm_argBool_sparc, 
   &dc_callvm_argChar_sparc, 
   &dc_callvm_argShort_sparc, 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_sparc64.c
--- a/dyncall/dyncall_callvm_sparc64.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_sparc64.c	Sat Mar 12 10:54:27 2016 +0300
@@ -145,6 +145,7 @@
   &dc_callvm_free_sparc64, 
   &dc_callvm_reset_sparc64, 
   &dc_callvm_mode_sparc64, 
+  NULL /* retStruct */,
   &dc_callvm_argBool_sparc64, 
   &dc_callvm_argChar_sparc64, 
   &dc_callvm_argShort_sparc64, 
@@ -174,6 +175,7 @@
   &dc_callvm_free_sparc64, 
   &dc_callvm_reset_sparc64, 
   &dc_callvm_mode_sparc64, 
+  NULL /* retStruct */,
   &dc_callvm_argBool_sparc64, 
   &dc_callvm_argChar_sparc64, 
   &dc_callvm_argShort_sparc64, 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_sparc_v9.c
--- a/dyncall/dyncall_callvm_sparc_v9.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_sparc_v9.c	Sat Mar 12 10:54:27 2016 +0300
@@ -103,6 +103,7 @@
   &dc_callvm_free_v9, 
   &dc_callvm_reset_v9, 
   &dc_callvm_mode_v9, 
+  NULL /* retStruct */,
   &dc_callvm_argBool_v9, 
   &dc_callvm_argChar_v9, 
   &dc_callvm_argShort_v9, 
@@ -132,6 +133,7 @@
   &dc_callvm_free_v9, 
   &dc_callvm_reset_v9, 
   &dc_callvm_mode_v9, 
+  NULL /* retStruct */,
   &dc_callvm_argBool_v9, 
   &dc_callvm_argChar_v9, 
   &dc_callvm_argShort_v9, 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_x64.c
--- a/dyncall/dyncall_callvm_x64.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_x64.c	Sat Mar 12 10:54:27 2016 +0300
@@ -31,7 +31,11 @@
 
 #include "dyncall_callvm_x64.h"
 #include "dyncall_alloc.h"
+#include "dyncall_signature.h"
 #include "dyncall_struct.h"
+#include <string.h>
+
+#define DC_UNIX_MEM_STRUCT_SIZE 16
 
 static DCCallVM* dc_callvm_new_x64(DCCallVM_vt* vt, DCsize size)
 {
@@ -47,6 +51,10 @@
   self->mRegCount.i = self->mRegCount.f =  0;
 
   dcVecInit(&self->mVecHead, size);
+#if defined(DC_UNUX)
+  self->mRetStruct = NULL;
+  self->mRetMem = NULL;
+#endif
   return (DCCallVM*)self;
 }
 
@@ -77,7 +85,6 @@
   }
 }
 
-
 static void dc_callvm_argLongLong_x64(DCCallVM* in_self, DClonglong x)
 {
   /* A long long always has 64 bits on the supported x64 platforms (lp64 on unix and llp64 on windows). */
@@ -88,6 +95,16 @@
     dcVecAppend(&self->mVecHead, &x, sizeof(DClonglong));
 }
 
+#if defined(DC_UNIX)
+static void dc_callvm_retStruct_x64(DCCallVM* in_self, DCstruct* s, DCpointer x)
+{
+  DCCallVM_x64* self = (DCCallVM_x64*)in_self;
+  self->mRetStruct = s;
+  self->mRetMem = x;
+  if (s->size > DC_UNIX_MEM_STRUCT_SIZE)
+    dc_callvm_argLongLong_x64(in_self, (DClonglong)x);
+}
+#endif
 
 static void dc_callvm_argBool_x64(DCCallVM* in_self, DCbool x)
 {
@@ -156,8 +173,66 @@
     dcVecAppend(&self->mVecHead, &x, sizeof(DCpointer));
 }
 
+// >= 0 n-th bit of bitmask means that n-th byte of structure is in INT field
+static int classify_struct_x64(DCstruct* s) {
+  DCsize i, j;
+  int bitmask = 0, submask;
+  for (i = 0; i < s->fieldCount; i++) {
+    DCfield *f = s->pFields + i;
+    for (j = 0; j < f->arrayLength; j++) {
+      int offset = f->offset + j*f->size;
+      switch (f->type) {
+      case DC_SIGCHAR_FLOAT:
+      case DC_SIGCHAR_DOUBLE:
+	break;
+      case DC_SIGCHAR_STRUCT:
+	submask = classify_struct_x64(f->pSubStruct);
+	bitmask |= (submask << offset);
+	break;
+      default:
+	submask = (1 << f->size) - 1;
+	bitmask  |= (submask << offset);
+      }
+    }
+  }
+  return bitmask;
+}
+
 static void dc_callvm_argStruct_x64(DCCallVM* in_self, DCstruct* s, DCpointer x)
 {
+  DCsize i, bitmask;
+  DCCallVM_x64* self = (DCCallVM_x64*)in_self;
+  /*printf("dc_callvm_argStruct_x64 size = %d\n", (int)s->size);@@@*/
+  if (s->size > DC_UNIX_MEM_STRUCT_SIZE) {
+    dcVecAppend(&self->mVecHead, x, s->size);
+    dcVecAlign(&self->mVecHead, sizeof(DClonglong));
+    return;
+  }
+  
+  // Avoid accessig bytes out of struct
+  DClonglong buf[2] = {0, 0};
+  memcpy(buf, x, s->size);
+  
+  bitmask = classify_struct_x64(s);
+  for (int i = 0; i*8 < s->size; i++) {
+    if (((bitmask >> (i*8)) & 0xff) != 0) {
+      // INTEGER class
+      if(self->mRegCount.i < numIntRegs)
+	self->mRegData.i[self->mRegCount.i++] = buf[i];
+      else
+	dcVecAppend(&self->mVecHead, &buf[i], sizeof(DClonglong));
+    } else {
+      // SSE class
+      if(self->mRegCount.f < numFloatRegs)
+	*(DClonglong*)&self->mRegData.f[self->mRegCount.f++] = buf[i];
+      else
+	dcVecAppend(&self->mVecHead, &buf[i], sizeof(DCdouble));
+    }
+  }
+}
+
+static void dc_callvm_argStruct_x64_old(DCCallVM* in_self, DCstruct* s, DCpointer x)
+{
   DCCallVM_x64* self = (DCCallVM_x64*)in_self;
   dcVecAppend(&self->mVecHead, x, s->size);
   /*printf("dc_callvm_argStruct_x64 size = %d\n", (int)s->size);@@@*/
@@ -167,7 +242,6 @@
   /*	  dcVecAppend(&self->mVecHead, &x, sizeof(DCpointer));@@@*/
 }
 
-
 /* Call. */
 void dc_callvm_call_x64(DCCallVM* in_self, DCpointer target)
 {
@@ -187,12 +261,60 @@
   );
 }
 
+struct STRUCT_FF { DCdouble f0; DCdouble f1; };
+struct STRUCT_FI { DCdouble f0; DClonglong f1; };
+struct STRUCT_IF { DClonglong f0; DCdouble f1; };
+struct STRUCT_II { DClonglong f0; DClonglong f1; };
+
+typedef struct STRUCT_FF (DCstructFFfunc)   (DCCallVM* vm,DCpointer funcptr);
+typedef struct STRUCT_FI (DCstructFIfunc)   (DCCallVM* vm,DCpointer funcptr);
+typedef struct STRUCT_IF (DCstructIFfunc)   (DCCallVM* vm,DCpointer funcptr);
+typedef struct STRUCT_II (DCstructIIfunc)   (DCCallVM* vm,DCpointer funcptr);
+
+#if defined(DC_UNIX)
+static void dc_callvm_callStruct_x64(DCCallVM* in_self, DCpointer target)
+{
+  DCsize bitmask;
+  DCCallVM_x64* self = (DCCallVM_x64*)in_self;
+  if (self->mRetStruct->size > DC_UNIX_MEM_STRUCT_SIZE) {
+    dc_callvm_call_x64(in_self, target);
+    return;
+  }
+  bitmask = classify_struct_x64(self->mRetStruct);
+  if ((bitmask & 0xff) != 0) {
+    if ((bitmask & 0xff00) != 0) {
+      DCstructIIfunc* fun = ((DCstructIIfunc*)&dc_callvm_call_x64);
+      struct STRUCT_II ret = (*fun)(in_self, target);
+      memcpy(self->mRetMem, &ret, self->mRetStruct->size);
+    } else {
+      DCstructIFfunc* fun = ((DCstructIFfunc*)&dc_callvm_call_x64);
+      struct STRUCT_IF ret = (*fun)(in_self, target);
+      memcpy(self->mRetMem, &ret, self->mRetStruct->size);
+    }
+  } else {
+    if ((bitmask & 0xff00) != 0) {
+      DCstructFIfunc* fun = ((DCstructFIfunc*)&dc_callvm_call_x64);
+      struct STRUCT_FI ret = (*fun)(in_self, target);
+      memcpy(self->mRetMem, &ret, self->mRetStruct->size);
+    } else {
+      DCstructFFfunc* fun = ((DCstructFFfunc*)&dc_callvm_call_x64);
+      struct STRUCT_FF ret = (*fun)(in_self, target);
+      memcpy(self->mRetMem, &ret, self->mRetStruct->size);
+    }
+  }
+}
+#endif
 
 DCCallVM_vt gVT_x64 =
 {
   &dc_callvm_free_x64
 , &dc_callvm_reset_x64
 , &dc_callvm_mode_x64
+#if defined(DC_UNIX)
+, dc_callvm_retStruct_x64
+#else
+, NULL /* retStruct */
+#endif
 , &dc_callvm_argBool_x64
 , &dc_callvm_argChar_x64
 , &dc_callvm_argShort_x64 
@@ -213,7 +335,11 @@
 , (DCfloatvmfunc*)      &dc_callvm_call_x64
 , (DCdoublevmfunc*)     &dc_callvm_call_x64
 , (DCpointervmfunc*)    &dc_callvm_call_x64
+#if defined(DC_UNIX)
+, dc_callvm_callStruct_x64
+#else
 , NULL /* callStruct */
+#endif
 };
 
 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_x64.h
--- a/dyncall/dyncall_callvm_x64.h	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_x64.h	Sat Mar 12 10:54:27 2016 +0300
@@ -102,6 +102,10 @@
   DCpointer      mpCallFunc;  /* Function to call.                                       */
   DCRegCount_x64 mRegCount;   /* Number of int/sse registers used for parameter passing. */
   DCRegData_x64  mRegData;    /* Parameters to be passed via registers.                  */
+#if defined(DC_UNIX) 
+  DCstruct*      mRetStruct;  /* Type of return structure.                               */
+  DCpointer      mRetMem;     /* Memory for return structure.                            */
+#endif
   DCVecHead      mVecHead;    /* Parameters to be pushed onto stack.                     */
 } DCCallVM_x64;
 
diff -r 33c333ec5ad2 dyncall/dyncall_callvm_x86.c
--- a/dyncall/dyncall_callvm_x86.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_callvm_x86.c	Sat Mar 12 10:54:27 2016 +0300
@@ -163,6 +163,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86
 , &dc_callvm_argChar_x86
 , &dc_callvm_argShort_x86 
@@ -208,6 +209,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86
 , &dc_callvm_argChar_x86
 , &dc_callvm_argShort_x86 
@@ -255,6 +257,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86
 , &dc_callvm_argChar_x86
 , &dc_callvm_argShort_x86
@@ -366,6 +369,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86_win32_fast
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86_win32_fast_ms
 , &dc_callvm_argChar_x86_win32_fast_ms
 , &dc_callvm_argShort_x86_win32_fast_ms
@@ -464,6 +468,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86_win32_fast
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86_win32_fast_gnu
 , &dc_callvm_argChar_x86_win32_fast_gnu
 , &dc_callvm_argShort_x86_win32_fast_gnu
@@ -509,6 +514,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86
 , &dc_callvm_argChar_x86
 , &dc_callvm_argShort_x86
@@ -553,6 +559,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86
 , &dc_callvm_argChar_x86
 , &dc_callvm_argShort_x86
@@ -581,6 +588,7 @@
   &dc_callvm_free_x86
 , &dc_callvm_reset_x86
 , &dc_callvm_mode_x86
+, NULL /* retStruct */
 , &dc_callvm_argBool_x86
 , &dc_callvm_argChar_x86
 , &dc_callvm_argShort_x86
diff -r 33c333ec5ad2 dyncall/dyncall_struct.c
--- a/dyncall/dyncall_struct.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncall/dyncall_struct.c	Sat Mar 12 10:54:27 2016 +0300
@@ -130,6 +130,7 @@
 	for (i = 0; i < s->fieldCount; i++) {
 		DCfield *f = s->pFields + i;
 		dcAlign(&s->size, f->alignment);
+		f->offset = s->size;
 		s->size += f->size;
 	}
 	dcAlign(&s->size, s->alignment);
diff -r 33c333ec5ad2 dyncallback/dyncall_args.h
--- a/dyncallback/dyncall_args.h	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncallback/dyncall_args.h	Sat Mar 12 10:54:27 2016 +0300
@@ -58,6 +58,7 @@
 DC_API DCfloat     dcbArgFloat    (DCArgs*);
 DC_API DCdouble    dcbArgDouble   (DCArgs*);
 DC_API DCpointer   dcbArgPointer  (DCArgs*);
+DC_API DCpointer   dcbArgStruct   (DCArgs*, DCstruct* s);
 
 #ifdef __cplusplus
 }
diff -r 33c333ec5ad2 dyncallback/dyncall_args_x64.c
--- a/dyncallback/dyncall_args_x64.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dyncallback/dyncall_args_x64.c	Sat Mar 12 10:54:27 2016 +0300
@@ -26,6 +26,8 @@
 
 
 #include "dyncall_args_x64.h"
+#include "dyncall_struct.h"
+#include <stdio.h>
 
 
 static int64* arg_i64(DCArgs* args)
@@ -69,6 +71,15 @@
 
 
 DCpointer   dcbArgPointer  (DCArgs* p) { return (DCpointer)   dcbArgLongLong(p); }
+DCpointer   dcbArgStruct   (DCArgs* p, DCstruct* s)
+{ 
+  if (s->size <= 16) {
+    printf("dcbArgStruct s->size=%d\n", s->size);
+  }
+  DCpointer r = p->stack_ptr;
+  p->stack_ptr += ((s->size + 7) >> 3); // align
+  return r;
+}
 
 DCdouble    dcbArgDouble   (DCArgs* p) { return *arg_f64(p); }
 DCfloat     dcbArgFloat    (DCArgs* p) { return *(float*)arg_f64(p); }
diff -r 33c333ec5ad2 dynload/dynload_syms_mach-o.c
--- a/dynload/dynload_syms_mach-o.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dynload/dynload_syms_mach-o.c	Sat Mar 12 10:54:27 2016 +0300
@@ -29,6 +29,7 @@
  dynamic symbol resolver for Mach-O
 
 */
+#include "../autovar/autovar_ARCH.h"
 
 #include "dynload.h"
 #include "dynload_alloc.h"
@@ -40,15 +41,16 @@
 
 #if defined(ARCH_X64) //@@@ use dyncall_macros.h
 #define MACH_HEADER_TYPE mach_header_64
+#define MACH_MAGIC MH_MAGIC_64
 #define SEGMENT_COMMAND segment_command_64
 #define NLIST_TYPE nlist_64
 #else
 #define MACH_HEADER_TYPE mach_header
+#define MACH_MAGIC MH_MAGIC
 #define SEGMENT_COMMAND segment_command
 #define NLIST_TYPE nlist
 #endif
 
-
 struct DLLib_
 {
 	char* libPath;
@@ -58,11 +60,139 @@
 
 struct DLSyms_
 {
+    const char** decompressedSymbols;
 	const char* pStringTable;
 	const struct NLIST_TYPE* pSymbolTable;
 	uint32_t symbolCount;
 };
 
+int isSameMacImageName(const char* libPath, const char* systemLibPath) {
+	if (!libPath || !systemLibPath)
+		return 0;
+	
+	if (!strcmp(libPath, systemLibPath))
+		return 1;
+	
+	if (strstr(libPath, "/libc.") && strstr(systemLibPath, "/libSystem."))
+		return 1;
+	
+	return 0;
+}
+
+///
+/// LEB128 DECODING (http://en.wikipedia.org/wiki/LEB128)
+///
+unsigned long long decode_leb128(const unsigned char** ppIn) {
+    const unsigned char* pIn = *ppIn;
+    unsigned char input;
+    unsigned long long decoded = 0;
+    int shift = 0;
+    do {
+        input = *pIn++;
+        decoded |= (((unsigned long long)(input & 0x7f)) << shift);
+        shift += 7;
+    } while (input & 0x80);
+    
+    *ppIn = pIn;
+    return decoded;
+}
+
+///
+/// DUMMY APPENDABLE STRING 
+///
+typedef struct _DCString {
+    char* buffer;
+    size_t capacity;
+} DCString;
+void initString(DCString* s, size_t capacity) {
+    if (!s)
+        return;
+    
+    s->buffer = dlAllocMem((s->capacity = capacity) + 1);
+    s->buffer[0] = '\0';
+}
+void appendString(DCString* s, const char* suffix) {
+    if (!s)
+        return;
+    
+    size_t suffixLen = strlen(suffix), sLen = strlen(s->buffer);
+    if (suffixLen + sLen > s->capacity) {
+        char* old = s->buffer;
+        initString(s, (suffixLen + sLen) * 3 / 2);
+        strcpy(s->buffer, old);
+        dlFreeMem(old);
+    }
+    strcpy(s->buffer + sLen, suffix);
+}
+void freeString(DCString* s) {
+	if (s->buffer) {
+		dlFreeMem(s->buffer);
+		s->buffer = NULL;
+		s->capacity = 0;
+	}
+}
+
+///
+/// MACHO COMPRESSED SYMBOLS TRIE VISITATION
+///
+typedef void (*SymbolOutputFunction)(const char* symbol, size_t index, void* payload);
+
+#define MAX_DEPTH 256
+
+size_t visitTrie(
+	const unsigned char* start, 
+	const unsigned char* current, 
+	const unsigned char* end, 
+	DCString* prefix, 
+	SymbolOutputFunction output, size_t indexOffset, void* payload, int depth) 
+{
+    size_t count = 0;
+	if (current < end && depth < MAX_DEPTH) {
+        unsigned char term_size = *current++;
+        if (term_size != 0) {
+            //const unsigned char* pos = current;
+            //unsigned long long flags = decode_leb128(&pos);
+            //unsigned long long addr = decode_leb128(&pos);
+            {
+				if (prefix && output)
+					output(prefix->buffer, indexOffset, payload);
+				
+				indexOffset++;
+				count++;
+			}
+            
+            current += term_size;
+        }
+
+        unsigned char iSubNode, nSubNodes = *current++;
+        
+        size_t prefixLen = prefix ? strlen(prefix->buffer) : 0;
+        for (iSubNode = 0; iSubNode < nSubNodes; iSubNode++) {
+            unsigned long long subNodeOffset;
+            const char* suffix = (const char*)current;
+            current += strlen(suffix) + 1;
+            if (!(subNodeOffset = decode_leb128(&current)))
+            	continue;
+            
+			appendString(prefix, suffix);
+			size_t subCount = visitTrie(start, start + subNodeOffset, end, prefix, output, indexOffset, payload, depth + 1);
+			count += subCount;
+			indexOffset += subCount;
+			if (prefix)
+				prefix->buffer[prefixLen] = '\0';
+        }
+    }
+    return count;
+}
+
+/// Callback to harvest symbols visited in trie
+void CopyToNthString(const char* str, size_t index, void* payload) {
+    const char** pstr = (const char**)payload;
+    size_t len = strlen(str);
+    char* copy = dlAllocMem(len + 1);
+    strcpy(copy, str);
+    pstr[index] = copy;
+}
 
 DLSyms* dlSymsInit(const char* libPath) 
 {
@@ -71,27 +201,59 @@
 	for (iImage = 0, nImages = _dyld_image_count(); iImage < nImages; iImage++)
 	{
 		const char* name = _dyld_get_image_name(iImage);
-		if (name && !strcmp(name, libPath))
+		if (isSameMacImageName(libPath, name))
 		{
 			const struct MACH_HEADER_TYPE* pHeader = (const struct MACH_HEADER_TYPE*) _dyld_get_image_header(iImage);
 			const char* pBase = ((const char*)pHeader);
+			if (!pHeader)
+				return NULL;
+
+			if (pHeader->magic != MACH_MAGIC)
+				return NULL;
 			if (pHeader->filetype != MH_DYLIB)
 				return NULL;
-			if (pHeader->flags & MH_SPLIT_SEGS)
+			if (pHeader->flags & MH_SPLIT_SEGS || !(pHeader->flags & MH_DYLDLINK))
 				return NULL;
-
-			if (pHeader)
+			
 			{
 				uint32_t iCmd, nCmds = pHeader->ncmds;
 				const struct load_command* cmd = (const struct load_command*)(pBase + sizeof(struct MACH_HEADER_TYPE));
 				
 				for (iCmd = 0; iCmd < nCmds; iCmd++) 
 				{
-					if (cmd->cmd == LC_SYMTAB) 
-					{
+					if (cmd->cmd == LC_DYLD_INFO_ONLY || cmd->cmd == LC_DYLD_INFO) {
+                        const struct dyld_info_command* dcmd = (const struct dyld_info_command*)cmd;
+                        const unsigned char* trie = (pBase + dcmd->export_off);
+                        
+                        pSyms = (DLSyms*)( dlAllocMem(sizeof(DLSyms)) );
+                        
+                        // First, get the number of symbols
+                        pSyms->symbolCount = visitTrie(trie, trie, trie + dcmd->export_size, NULL, NULL, 0, NULL, 0);
+                        
+                        if (pSyms->symbolCount) {
+                            // Now revisit and copy symbols to their destination
+                            DCString s;
+                            size_t decompSize = pSyms->symbolCount * sizeof(char*) * 2;
+                            pSyms->decompressedSymbols = dlAllocMem(decompSize);
+                            memset(pSyms->decompressedSymbols, decompSize, 0);
+                            initString(&s, 1024);
+                            visitTrie(trie, trie, trie + dcmd->export_size, &s, CopyToNthString, 0, pSyms->decompressedSymbols, 0);
+                            freeString(&s);
+                        }
+                        
+                        return pSyms;
+                    }
+                    
+                    if (cmd->cmd & LC_REQ_DYLD) {
+                    	return NULL; // "unknown load command required for execution";
+                    }
+                    if (cmd->cmd == LC_SYMTAB) 
+                    {
 						const struct symtab_command* scmd = (const struct symtab_command*)cmd;
 					
 						pSyms = (DLSyms*)( dlAllocMem(sizeof(DLSyms)) );
+						//memset(pSyms, sizeof(DLSyms), 0);
+						pSyms->decompressedSymbols = NULL;
 						pSyms->symbolCount = scmd->nsyms;
 						pSyms->pStringTable = pBase + scmd->stroff;
 						pSyms->pSymbolTable = (struct NLIST_TYPE*)(pBase + scmd->symoff);
@@ -101,7 +263,6 @@
 					cmd = (const struct load_command*)(((char*)cmd) + cmd->cmdsize);
 				}
 			}
-			break;
 		}
 	}
 	return NULL;
@@ -113,6 +274,18 @@
 	if (!pSyms)
 		return;
 	
+    if (pSyms->decompressedSymbols) {
+        size_t i, n = pSyms->symbolCount;
+        for (i = 0; i < n; i++) {
+            const char* str = pSyms->decompressedSymbols[i];
+            if (str) {
+            	dlFreeMem(str);
+            	pSyms->decompressedSymbols[i] = NULL;
+            }
+        }
+        dlFreeMem(pSyms->decompressedSymbols);
+        pSyms->decompressedSymbols = NULL;
+    }
 	dlFreeMem(pSyms);
 }
 
@@ -133,6 +306,11 @@
 	if (nl->n_un.n_strx <= 1)
 		return NULL; // would be empty string anyway
 	
+    if (!(nl->n_type & N_EXT) || (nl->n_type & N_PEXT))
+        return NULL; // not a public external symbol
+    if (nl->n_type & N_STAB)
+        return NULL;
+    
 	//TODO skip more symbols based on nl->n_desc and nl->n_type ?
 	return nl;
 }
@@ -140,6 +318,12 @@
 
 const char* dlSymsName(DLSyms* pSyms, int index)
 {
+    if (!pSyms || index >= pSyms->symbolCount)
+        return NULL;
+ 
+    if (pSyms->decompressedSymbols)
+        return pSyms->decompressedSymbols[index];
+            
 	const struct NLIST_TYPE* nl = get_nlist(pSyms, index);
 	if (!nl)
 		return NULL;
@@ -150,6 +334,12 @@
 
 void* dlSymsValue(DLSyms* pSyms, int index)
 {
+    if (!pSyms || index >= pSyms->symbolCount)
+        return NULL;
+    
+    if (pSyms->decompressedSymbols)
+        return NULL; // TODO use dlFindSymbol
+            
 	const struct NLIST_TYPE* nl = get_nlist(pSyms, index);
 	if (!nl)
 		return NULL;
diff -r 33c333ec5ad2 dynload/dynload_unix.c
--- a/dynload/dynload_unix.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dynload/dynload_unix.c	Sat Mar 12 10:54:27 2016 +0300
@@ -41,7 +41,7 @@
 
 DLLib* dlLoadLibrary(const char* libPath)
 {
-  return (DLLib*)dlopen(libPath,RTLD_NOW|RTLD_GLOBAL);
+  return (DLLib*)dlopen(libPath,RTLD_LAZY|RTLD_GLOBAL);
 }
 
 
diff -r 33c333ec5ad2 dynload/dynload_windows.c
--- a/dynload/dynload_windows.c	Sun Mar 06 23:30:44 2016 +0100
+++ b/dynload/dynload_windows.c	Sat Mar 12 10:54:27 2016 +0300
@@ -41,7 +41,9 @@
 DLLib* dlLoadLibrary(const char* libPath)
 {
   if (libPath != NULL) {
-    return (DLLib*) LoadLibraryA(libPath);
+    return (DLLib*) LoadLibraryExW((wchar_t*)(void*)libPath, NULL, LOAD_WITH_ALTERED_SEARCH_PATH);
+  	// http://msdn.microsoft.com/en-us/library/windows/desktop/ms684179(v=vs.85).aspx
+    // return (DLLib*) LoadLibraryExA(libPath, NULL, LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR | LOAD_LIBRARY_DEFAULT_DIRS);
     /*return (DLLib*) LoadLibraryA(libPath, NULL, LOAD_WITH_ALTERED_SEARCH_PATH);*/
   } else {
     return (DLLib*) GetModuleHandle(NULL);
